pypit/arextract.py

```diff
@@ -451,10 +466,13 @@ def optimal_extract(slf, det, specobjs, sciframe, varframe,
     cr_mask = 1.0-crmask
     # Object model image
     obj_model = np.zeros_like(varframe)
+    gdslits = np.where(~slf._maskslits[det-1])[0]
     # Loop on slits
     for sl in range(len(specobjs)):
+        if sl not in gdslits:
+            continue
         # Loop on objects
-        nobj = scitrace[sl]['traces'].shape[1]
+        nobj = scitrace[sl]['nobj']
         for o in range(nobj):
```

Comment:
Not that it makes a big difference, but a suggestion:
```
if nobj == 0:
    continue
```

Reply:
added